include/SDL3/SDL_ioasync.h

```diff
@@ -0,0 +1,500 @@
+ *
+ * And, of course, to match the synchronous SDL_LoadFile, we offer
+ * SDL_LoadFileAsync as a convenience function. This will handle allocating
+ * a buffer, slurping in the file data, and null-terminating it; you still
+ * get a task handle to check later or add to a task group.
+ */
+
+#ifndef SDL_ioasync_h_
+#define SDL_ioasync_h_
+
+#include <SDL3/SDL_stdinc.h>
+
+#include <SDL3/SDL_begin_code.h>
+/* Set up for C function definitions, even when using C++ */
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * The asynchronous I/O operation structure.
+ *
+ * This operates as an opaque handle. One can then request read or write
+ * operations on it.
+ *
+ * \since This struct is available since SDL 3.0.0.
+ *
+ * \sa SDL_IOAsyncFromFile
+ */
+typedef struct SDL_IOAsync SDL_IOAsync;
+
+/**
+ * Types of asynchronous I/O tasks.
+ *
+ * \since This enum is available since SDL 3.0.0.
+ */
+typedef enum SDL_IOAsyncTaskType
+{
+    SDL_IOASYNC_TASK_READ,   /**< A read operation. */
+    SDL_IOASYNC_TASK_WRITE,  /**< A write operation. */
+    SDL_IOASYNC_TASK_CLOSE   /**< A close operation. */
+} SDL_IOAsyncTaskType;
+
+/**
+ * Possible outcomes of an asynchronous I/O task.
+ *
+ * \since This enum is available since SDL 3.0.0.
+ */
+typedef enum SDL_IOAsyncResult
+{
+    SDL_IOASYNC_COMPLETE,  /**< request was completed without error */
+    SDL_IOASYNC_FAILURE,   /**< request failed for some reason; check SDL_GetError()! */
+    SDL_IOASYNC_CANCELLED  /**< request was cancelled before completing. */
+} SDL_IOAsyncResult;
+
+/**
+ * Information about a completed asynchronous I/O request.
+ *
+ * \since This struct is available since SDL 3.0.0.
+ */
+typedef struct SDL_IOAsyncOutcome
+{
+    SDL_IOAsync *ioasync;   /**< what generated this task. This pointer will be invalid if it was closed! */
+    SDL_IOAsyncTaskType type;  /**< What sort of task was this? Read, write, etc? */
+    SDL_IOAsyncResult result;  /**< the result of the work (success, failure, cancellation). */
+    void *buffer;  /**< buffer where data was read/written. */
+    Uint64 offset;  /**< offset in the SDL_IOAsync where data was read/written. */
+    Uint64 bytes_requested;  /**< number of bytes the task was to read/write. */
+    Uint64 bytes_transferred;  /**< actual number of bytes that were read/written. */
+    void *userdata;    /**< pointer provided by the app when starting the task */
+} SDL_IOAsyncOutcome;
+
+/**
+ * An opaque handle for asynchronous I/O tasks.
+ *
+ * Each asynchronous read or write operation generates a task, which will
+ * complete at some time in the future. This handle is used to track the
+ * progress of that task.
+ *
+ * Tasks can be queried for completion individually, or in a group using
+ * an SDL_IOAsyncGroup.
+ *
+ * \since This struct is available since SDL 3.0.0.
+ *
+ * \sa SDL_ReadIOAsync
+ * \sa SDL_WriteIOAsync
+ */
+typedef struct SDL_IOAsyncTask SDL_IOAsyncTask;
+
+/**
+ * A logical grouping of SDL_IOAsyncTask handles.
```

Comment:
The description here doesn't make it immediately obvious to me that the set of tasks in the group can constantly change. I think part of the problem is the name "group" which makes me think I'm statically taking a set of tasks and grouping them together, rather than using this object as a mechanism to submit tasks.

What if you named it SDL_IOWorkQueue?

Reply:
Calling it a queue is more appropriate IMO too. Though the purpose for the queue is really more about the completions than the work itself. Maybe SDL_IOCompletionQueue or just SDL_IOQueue?

Reply:
SDL_AsyncIOTask, SDL_AsyncIOQueue? That makes a certain sense to me.

Reply:
In 5829b1ff4fb0b9859c0f46e3eb0dc7f4ce5f5b7a, I just went with SDL_IOAsyncQueue.